### PR TITLE
Feat/dsg-903/formConfigValidation

### PR DESCRIPTION
**Pull request by Ortega-Dan**
[DSG-903: Form Configuration validation](https://dol-ewf.atlassian.net/browse/DSG-903)

## Description

Ability to identify issues with FormConfigurations when saved, to avoid having data validation issues during customer data submit due to not schema-configured components.

## How Has This Been Tested?

Unit tests and API testing.

## Types of changes

Patch
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation change (documentation only, will not require redeployment)
- [ ] Refactor or cleanup (functionality unchanged)
Minor
- [x] New feature (non-breaking change which adds functionality)
Major
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

- [x] I have confirmed all acceptance criteria defined by the story have been met

### CODE REVIEW

File: service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java [hunk 1]
```diff
@@ -70,9 +84,123 @@ public FormConfiguration saveFormConfiguration(final FormConfiguration formConfi
                         formConfiguration.getFormConfigurationKey())
                 .ifPresent(existing -> formConfiguration.setId(existing.getId()));
 
+        validateFormConfiguration(formConfiguration);
+
         return repository.save(formConfiguration);
     }
 
+    private void validateFormConfiguration(FormConfiguration formConfiguration) {
```
Comment (asultan001) [inline service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java:92]: This validation looks good to me but the messsages make it a bit hard to know which components are invalid or incorrect. Could we provide further context on the messages to help troubleshoot and fix the schema definition or the components that are invalid?

File: service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java [hunk 2]
```diff
@@ -70,9 +84,123 @@ public FormConfiguration saveFormConfiguration(final FormConfiguration formConfi
...
+    }
+
+    private void validateParentComponentConfig(
+            NuvalenceFormioComponent parentConfig,
+            String schemaKey,
+            Set<String> keyUniquenessSet,
+            List<NuvalenceFormioValidationExItem> validationErrors) {
+
+        if (schemaKey == null) {
+            addConfigErrorMessage(parentConfig.getKey(), "Null schema key", validationErrors);
```
Comment (asultan001) [inline service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java:124]: Can we provide more context to help in troubleshooting/fixing this? Like what component was missing the schema key or something that would be helpful here.

File: service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java [hunk 3]
```diff
@@ -70,9 +84,123 @@ public FormConfiguration saveFormConfiguration(final FormConfiguration formConfi
...
+            return;
+        }
+
+        Optional<Schema> schemaOptional = schemaService.getSchemaByKey(schemaKey);
+        if (schemaOptional.isEmpty()) {
+            addConfigErrorMessage(
+                    parentConfig.getKey(),
+                    "Schema with key \"" + schemaKey + "\" not found.",
```
Comment (asultan001) [inline service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java:132]: Same here, can we say what component this is in reference to?

File: service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java [hunk 4]
```diff
@@ -70,9 +84,123 @@ public FormConfiguration saveFormConfiguration(final FormConfiguration formConfi
...
+        }
+
+        Schema schema = schemaOptional.get();
+
+        if (schema.getDynaProperties().length > 0
+                && (parentConfig.getComponents() == null
+                        || parentConfig.getComponents().isEmpty())) {
+            addConfigErrorMessage(
+                    parentConfig.getKey(),
+                    "Component is a parent in schema but is missing its children",
```
Comment (asultan001) [inline service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java:144]: Would be helpful to know what component this is in reference to.

File: service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java [hunk 5]
```diff
@@ -70,9 +84,123 @@ public FormConfiguration saveFormConfiguration(final FormConfiguration formConfi
...
+    private void validateChildComponentConfig(
+            NuvalenceFormioComponent childConfig,
+            Schema schema,
+            Set<String> keyUniquenessSet,
+            List<NuvalenceFormioValidationExItem> validationErrors) {
+
+        if (!childConfig.getProps().isDataBacked()) {
+            return;
+        } else if (childConfig.getKey() == null) {
+            addConfigErrorMessage(null, "A component is missing its key", validationErrors);
```
Comment (asultan001) [inline service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java:165]: Which component is missing it's key?

File: service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java [hunk 6]
```diff
@@ -70,9 +84,123 @@ public FormConfiguration saveFormConfiguration(final FormConfiguration formConfi
...
+            return;
+        } else if (!keyUniquenessSet.add(childConfig.getKey())) {
+            addConfigErrorMessage(
+                    childConfig.getKey(),
+                    "Component key is found more than once in the form configuration",
```
Comment (asultan001) [inline service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java:170]: what component is this about? if we know which components have duplicates can we call them out?

File: service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java [hunk 7]
```diff
@@ -70,9 +84,123 @@ public FormConfiguration saveFormConfiguration(final FormConfiguration formConfi
...
+                    validationErrors);
+            return;
+        }
+
+        String[] keyParts = childConfig.getKey().split("\\.");
+        DynaProperty schemaProperty = schema.getDynaProperty(keyParts[keyParts.length - 1]);
+        if (schemaProperty == null) {
+            addConfigErrorMessage(
+                    childConfig.getKey(),
+                    "Component key not found in schema with key '" + schema.getKey() + "'",
```
Comment (asultan001) [inline service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java:180]: what component and what key?

File: service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java (hunk 1 above)
Comment (Ortega-Dan) [reply service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java:92]: Thank you Abe,
I do not have a way to debug it right now in order to show it. But the specific key of the component is in the controlName property, that is set here:
https://github.com/Nuvalence/dsgov-work-manager/blob/40bb659aa76b1847e8cf37a66345f53276404000/service/src/main/java/io/nuvalence/workmanager/service/config/exceptions/model/NuvalenceFormioValidationExItem.java#L12-L12

and is included in each error message construction:
https://github.com/Nuvalence/dsgov-work-manager/blob/40bb659aa76b1847e8cf37a66345f53276404000/service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java#L192-L202
and
https://github.com/Nuvalence/dsgov-work-manager/blob/40bb659aa76b1847e8cf37a66345f53276404000/service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java#L178-L181

Comment (Ortega-Dan) [reply service/src/main/java/io/nuvalence/workmanager/service/service/FormConfigurationService.java:92]: After making this comment, I anyway realize the names could be better for those specific variables. So just changed them.

**Review by asultan001 — APPROVED** [review]
Got it, can you include a few example api responses as a jira comment to the story? It would help see the full response.

**Comment by Ortega-Dan** [comment]
> Got it, can you include a few example api responses as a jira comment to the story? It would help see the full response.

Thx Abe, sure, I'll merge in order to provide the response since not yet done with my personal local env.

**Pull request merged**